javascript/react-js/react-testing-part-two.md

```diff
@@ -0,0 +1,160 @@
+
+  return (
+    <label htmlFor={id}>
+      What is your favorite wild animal?
+      <input id={id} onChange={inputHandler} />
+    </label>
+  );
+};
+
+export default FavoriteInput;
+~~~
+
+Nothing fancy. `FavoriteInput` is a simple component with a couple props passed in. We're interested in the `onChange` prop. We have no idea what the function does. We have no idea how the function will affect the application. All we know is it must be called when user types in the input box. Let's test it. 
+
+~~~javascript
+// FavoriteInput.test.js
+
+import React from "react";
+import { render, screen } from "@testing-library/react";
+import "@testing-library/jest-dom";
+import userEvent from "@testing-library/user-event";
+import FavoriteInput from "./FavoriteInput";
+
+describe("Favorite Input", () => {
+  it("calls onChange correct number of times", () => {
+    const onChangeMock = jest.fn();
+    render(<FavoriteInput onChange={onChangeMock} />);
+    const input = screen.getByRole("textbox");
+
+    userEvent.type(input, "Lion");
+
+    expect(onChangeMock).toHaveBeenCalledTimes(4);
+  });
+
+  it("calls onChange with correct argument(s) on each input", () => {
+    const onChangeMock = jest.fn();
+    render(<FavoriteInput onChange={onChangeMock} />);
+    const input = screen.getByRole("textbox");
+
+    userEvent.type(input, "Ox");
+
+    expect(onChangeMock).toHaveBeenNthCalledWith(1, "O");
+    expect(onChangeMock).toHaveBeenNthCalledWith(2, "Ox");
+  });
+
+  it("input has correct values", () => {
+    const onChangeMock = jest.fn();
+    render(<FavoriteInput onChange={onChangeMock} />);
+    const input = screen.getByRole("textbox");
+
+    userEvent.type(input, "Whale");
+
+    expect(input).toHaveValue("Whale");
+  });
+});
+~~~
+
+Three tests and we are done with this component. Take some time to figure out what functions come from which package. 
+We mock the `onChange` handler using one of jest features, `jest.fn()`. For the first test, we assert that the mock function is invoked correct number of times. While the second test ensures that the mock function is called with the correct arguments. The third test seems redundant, and it is; it's just here to show what other ways we could've tested the component.   
+
+But what if you want to set up your mocks in a `beforeEach` block rather than in every test? That's fine in some cases. Though, having all of the setup for a test in the same block as the test itself makes it easier to understand any particular test as it eliminates the need to check the whole file for context. This makes the reviewing of subsequent changes in a project down the road substantially easier. Additionally, it decreases the chance of having leakage create problems throughout the test suite. Unless your test file is getting real long and the test prep itself is dozens of lines in length, default to setting up in each test case; otherwise, you may use `beforeEach`.
```

Comment:
```suggestion
But what if you want to set up your mocks in a `beforeEach` block rather than in every test? That's fine in some cases. Though, having all of the setup for a test in the same block as the test itself makes it easier to understand any particular test as it eliminates the need to check the whole file for context. This makes the reviewing of subsequent changes in a project down the road substantially easier. Additionally, it decreases the chance of having leakage create problems throughout the test suite. Unless your test file is getting really long and the test prep itself is dozens of lines in length, default to setting up in each test case; otherwise, you may use `beforeEach`.
```